iwyu_output.cc

```diff
@@ -605,6 +619,29 @@ void IwyuFileInfo::ReportForwardDeclareUse(SourceLocation use_loc,
   LogSymbolUse("Marked fwd-decl use of decl", symbol_uses_.back());
 }
 
+void IwyuFileInfo::ReportUsingDeclUse(SourceLocation use_loc,
+                                      const UsingDecl* using_decl,
+                                      bool in_cxx_method_body,
+                                      const char* comment) {  
+
+  // If accessing a symbol through a using decl in the same file that contains
+  // the using decl, we must mark the using decl as referenced. At the end of
+  // traversing the AST, we check to see if a using decl is unreferenced and a
+  // full use of one of its shadow decls so that the source file continues to
+  // compile.
+  typename map<const UsingDecl*, bool>::iterator using_decl_status = 
+    using_decl_referenced_.find(using_decl);
+
+  if(using_decl_status != using_decl_referenced_.end()) {
+    using_decl_status->second = true;
+  }
+
+  // When a symbol is accessed through a using decl, we must report
+  // that as a full use of the using decl because whatever file that
+  // using decl is in is now required.
+  ReportFullSymbolUse(use_loc, using_decl, in_cxx_method_body, comment);
```

Comment:
This is an interesting thing to consider because I think the difference in functionality is subtle. In the end, the correct includes will be reported either way, but things like logs and error messages might be slightly different. One difference already seems to be that the decls are reported as they appear in source code.  Consider the following;

```cpp
using std::swap;
std::vector a, b;
swap(a, b);
```

If we report swap as being used through the using decl, as you suggest, it is reported as std::swap.  If we are lazy about it and let the AST traversal handle it, it's reported as simply swap.  I'm not certain which one we ultimately want.  Previous functionality (before my change) is to report as std::swap because the shadow decls were added greedily.

I will experiment with this to see what else is different.